anomalib/models/cfa/torch_model.py

```diff
@@ -0,0 +1,387 @@
+
+        Args:
+            distance (Tensor): Distance computed using target oriented features.
+
+        Returns:
+            Tensor: CFA loss.
+        """
+        num_neighbors = self.num_nearest_neighbors + self.num_hard_negative_features
+        distance = distance.topk(num_neighbors, largest=False).values
+
+        score = distance[:, :, : self.num_nearest_neighbors] - (self.radius**2).to(distance.device)
+        l_att = torch.mean(torch.max(torch.zeros_like(score), score))
+
+        score = (self.radius**2).to(distance.device) - distance[:, :, self.num_hard_negative_features :]
+        l_rep = torch.mean(torch.max(torch.zeros_like(score), score - 0.1))
+
+        loss = (l_att + l_rep) * 1000
+
+        return loss
+
+    def compute_score(self, distance: Tensor) -> Tensor:
+        """Compute score based on the distance.
+
+        Args:
+            distance (Tensor): Distance tensor computed using target oriented
+                features.
+
+        Returns:
+            Tensor: Score value.
+        """
+        distance = torch.sqrt(distance)
+
+        n_neighbors = self.num_nearest_neighbors
+        distance = distance.topk(n_neighbors, largest=False).values
+
+        distance = (F.softmin(distance, dim=-1)[:, :, 0]) * distance[:, :, 0]
+        distance = distance.unsqueeze(-1)
+
+        score = rearrange(distance, "b (h w) c -> b c h w", h=self.scale[0], w=self.scale[1])
+        return score.detach()
+
+    def compute_anomaly_map(self, score: Tensor) -> Tensor:
+        """Compute anomaly map based on the score.
+
+        Args:
+            score (Tensor): Score tensor.
+
+        Returns:
+            Tensor: Anomaly map.
+        """
+        anomaly_map = score.mean(dim=1, keepdim=True)
+        anomaly_map = F.interpolate(anomaly_map, size=self.input_size, mode="bilinear", align_corners=False)
+
+        gaussian_blur = GaussianBlur2d(sigma=4).to(score.device)
+        anomaly_map = gaussian_blur(anomaly_map)  # pylint: disable=not-callable
+        return anomaly_map
+
+    def forward(self, input_tensor: Tensor) -> Tensor:
+        """Forward pass.
+
+        Args:
+            input_tensor (Tensor): Input tensor.
+
+        Raises:
+            ValueError: When the memory bank is not initialized.
+
+        Returns:
+            Tensor: Loss or anomaly map depending on the train/eval mode.
+        """
+        if self.memory_bank.ndim == 0:
+            raise ValueError("Memory bank is not initialized. Run `initialize_centroid` method first.")
+
+        self.feature_extractor.eval()
+        with torch.no_grad():
+            features = self.feature_extractor(input_tensor)
+            features = list(features.values())
+
+        if hasattr(self, "scale") is False:
+            self.scale = features[0].size()[-2:]
+
+        target_features = self.descriptor(features)
+        distance = self.compute_distance(target_features)
+
+        if self.training:
+            output = self.compute_loss(distance)
```

Comment:
In our other models we leave the loss computation to the lightning model. Probably best to do the same here for consistency. In this case we could define a `CfaLoss` class which is called from the `training_step` method of the lightning model.